src/cli.js

```diff
@@ -24,6 +24,8 @@ Commands:
 
 Options:
   -o, --out [file]         Output directory for build (defaults to dist)
+  
+  --esm                    Create an ES Module build output
```

Comment:
I think we can omit this flag and instead detect automatically.

For example:

```js
const esm = buildFile.endsWith('.mjs') ||
  (buildFile.endsWith('.js') && require('package.json').type === 'module')
```

Reply:
It might be a bit magic to do this reverse detection of the package type perhaps? Perhaps we leave the flag as well and treat this as a default though...

Reply:
The flag doesn't really make sense in my option because its not really an option. The entrypoint either is or it isn't ESM.

When I run `node index.mjs` then it "just works" so I would expect `ncc run index.mjs` to just work as well.

Reply:
The thing is that the authoring format and the consumption format don't have to be the same. Also we don't currently allow customizing the file name.

Another option could even be to output ES modules by default...

Reply:
> the authoring format and the consumption format don't have to be the same

I don't think that is the goal of `ncc`. The number 1 design goal is "Zero configuration" from the README.

We need to follow Node.js conventions here, especially because publish with `npm` will use the default `package.json` which might contain `type: module`.

Reply:
Right now, ncc will convert ES modules into CommonJS for publishing. Changing this default is a breaking change.

Reply:
Yep, we can note it as a breaking change

Reply:
Ok, we can do that. Let's try to align it with this webpack PR release then as well - https://github.com/webpack/webpack/pull/13622.

Are there any other breaking changes we should make at the same time? I will aim to update the PR this week.

Reply:
I can't think of any other breaking changes we need right now

Reply:
Two possible issues with making the major change initially instead of having an `--esm` flag to start:
1. Updating the self build and all the tests is quite a bit of work
2. The v8cache can't be supported for --esm at all unfortunately.
Let me know if you still want to go ahead with this or use the flag.